exploration/variable-mutability.md

```diff
@@ -0,0 +1,244 @@
+		<dt>Pull Request</dt>
+		<dd>#000</dd>
+	</dl>
+</details>
+
+## Objective
+
+_What is this proposal trying to achieve?_
+
+Describe how variables are named and how externally passed variables
+and internally defined variables interact.
+
+## Background
+
+_What context is helpful to understand this proposal?_
+
+- [Issue 310](https://github.com/unicode-org/message-format-wg/issues/310)
+
+The term **local variable** refers to a value that is defined in a declaration.
+
+The term **external variable** refers to a value that is passed to the
+message formatter by name and which can be referred to in an expression.
+
+## Use-Cases
+
+_What use-cases do we see? Ideally, quote concrete examples._
+
+- Users want to reference external variables in expressions.
+- Users can modify external variables using declarations.
+  For example, they can perform a text transformation or assign reusable formatting options.
+  > ```
+  > let $foo = {$bar :uppercase}
+  > let $baz = {$someNumber :number groupingUsed=false}
+  > ```
+- Users, such as translators, want to annotate a variable
+  (either local or external) without invalidating
+  existing use of the variable in pattern strings. For example:
+  > ```
+  > let $foo = {$foo :transform}
+  > match {$a :plural} {$b :plural}
+  > when 0   0   {...{$foo}...}
+  > when 0   one {...{$foo}...}
+  > when 0   *   {...{$foo}...}
+  > when one 0   {...{$foo}...}
+  > when one one {...{$foo}...}
+  > when one *   {...{$foo}...}
+  > when *   0   {...{$foo}...}
+  > when *   one {...{$foo}...}
+  > when *   *   {...{$foo}...}
+  > ```
+- Users want to perform multiple transforms on a value.
+  Since our syntax does not permit embedding or chaining, this requires multiple declarations.
+  > ```
+  > let $foo = {$foo :text-transform transform=uppercase}
+  > let $foo = {$foo :trim}
+  > let $foo = {$foo :sanitize target=html}
+  > ```
+  >
+  > This can also be achieved by renaming:
+  >
+  > ```
+  > let $foo1 = {$foo :text-transform transform=uppercase}
+  > let $foo2 = {$foo1 :trim}
+  > let $foo3 = {$foo2 :sanitize target=html}
+  > ```
+- Users want to impose typing on (we say "annotate") external variables
+  or literals:
+  > ```
+  > let $fooAsNumber = {$foo :number}
+  > let $anotherNumber = {42 :number}
+  > ```
+- Users may wish to provide complex annotations which are reused across mulitple patterns
+
+  > ```
+  > let $count = {$count :number}
+  > let $date = {$date :datetime dateStyle=long}
+  > match {$count}
+  > when 1 {You received one message on {$date}}
+  > when * {You received {$count} messages on {$date}}
+  > ```
+
+- Implementers need to know what value is associated with a named variable, see #299.
+
+## Requirements
+
+_What properties does the solution have to manifest to enable the use-cases above?_
+
+- Be able to re-annotate variables without having to rename them in the message body
+- Allow static analysis to detect mistakes when referencing an undefined local variable
```

Comment:
```suggestion
```

I don't think this should be considered as a requirement. At best it's a nice-to-have feature improving static analysis when a single translated message needs to be considered in isolation, and not with respect to its form in the source locale.

Reply:
Would you prefer a use case for this? This specific list was taken from @stam's comment in #310 near the end (after I had written a longer and more detailed set of requirements). The "static analysis" requirement is kind of "boiling down" the question of whether people will make mistakes with overwriting.